### PR TITLE
Avoid double-type-checks in CSharpSemanticModel

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

**Comment by alrz** [comment]
If it's preferable to do this in other places in one go, I have an unpushed commit for that.

**Comment by alrz** [comment]
Is it preferable to do this in other places too? (`is` pattern is slightly more optimal since it emits `dup` rather than loading the field twice).

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 1]
```diff
@@ -1758,94 +1748,92 @@ private bool IsInTypeofExpression(int position)
...
+                // Empty typically indicates an error symbol that was created because no real
+                // symbol actually existed.
+                return SymbolInfoFactory.Create(ImmutableArray<Symbol>.Empty, LookupResultKind.Empty, isDynamic);
+            }
+            else
+            {
+                // Caas clients don't want ErrorTypeSymbol in the symbols, but the best guess
+                // instead. If no best guess, then nothing is returned.
+                var builder = ArrayBuilder<Symbol>.GetInstance();
+                foreach (var s in symbols)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: FYI.  Compiler style is to be explicit about the type *unless* it is clear from the initializer.  So in this case, keep as ```Symbol symbol```

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler  for reviews here.

Note @alrz people are in crunch mode for 15.3.  So it may take a while for this to get some eyes.

**Comment by CyrusNajmabadi** [comment]
> Is it preferable to do this in other places too

Feel free to do for the IDE side (i'm happy to review those).  You should get info from some people from the compiler team if you should continue doing this on the compiler side.  That said, this seems like a pure win.  Doing a single check instead of is+cast is definitely preferable.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 2]
```diff
@@ -284,23 +283,19 @@ private BoundExpression GetSpeculativelyBoundExpression(int position, Expression
             }
             else if (binder.InCref)
             {
-                if (expression.IsKind(SyntaxKind.QualifiedName))
+                if (expression is QualifiedNameSyntax qualified)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Kind checks are preferable to type checks, where possible.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 3]
```diff
@@ -4868,12 +4768,12 @@ private static NamespaceOrTypeSymbol ToLanguageSpecific(INamespaceOrTypeSymbol c
                 return null;
             }
 
-            var result = container as NamespaceOrTypeSymbol;
-            if ((object)result == null)
+            if (!(container is NamespaceOrTypeSymbol namespaceOrTypeSymbol))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Nit: Consider swapping this condition around (using positive test instead of negative).

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: On second thought, don't bother. This is a recurring pattern in other methods anyways.

---
In reply to: [121507519](https://github.com/dotnet/roslyn/pull/19941#discussion_r121507519) [](ancestors = 121507519)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 4]
```diff
@@ -18,7 +17,7 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// <summary>
     /// Allows asking semantic questions about a tree of syntax nodes in a Compilation. Typically,
     /// an instance is obtained by a call to <see cref="Compilation"/>.<see
-    /// cref="Compilation.GetSemanticModel"/>. 
+    /// cref="Compilation.GetSemanticModel"/>.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >/// cref="Compilation.GetSemanticModel"/>. [](start = 4, length = 42)

Please revert unnecessary whitespace changes.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 5]
```diff
@@ -4676,54 +4597,36 @@ protected sealed override TypeInfo GetTypeInfoCore(SyntaxNode node, Cancellation
...
         {
-            var nameSyntax = node as IdentifierNameSyntax;
-            return nameSyntax == null ? PreprocessingSymbolInfo.None : GetPreprocessingSymbolInfo(nameSyntax);
+            return node is IdentifierNameSyntax nameSyntax
+                ? GetPreprocessingSymbolInfo(nameSyntax)
+                : PreprocessingSymbolInfo.None;
         }
 
-        protected sealed override ISymbol GetDeclaredSymbolCore(SyntaxNode declaration, CancellationToken cancellationToken)
+        protected sealed override ISymbol GetDeclaredSymbolCore(SyntaxNode node, CancellationToken cancellationToken)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Wouldn't it better if we run the second switch (on kind) first?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 6]
```diff
@@ -4676,54 +4597,36 @@ protected sealed override TypeInfo GetTypeInfoCore(SyntaxNode node, Cancellation
...
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            var node = (CSharpSyntaxNode)declaration;
-
-            var accessor = node as AccessorDeclarationSyntax;
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: AccessorDeclarationSyntax is sealed, so I think it should be moved to the next switch (on kind).

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 4 above)
Comment (alrz) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Done.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 7]
```diff
@@ -291,17 +291,14 @@ private BoundExpression GetSpeculativelyBoundExpression(int position, Expression
...
                 {
-                    var typeSyntax = expression as TypeSyntax;
-                    if (typeSyntax != null)
-                    {
-                        var crefWrapper = typeSyntax is PredefinedTypeSyntax ?
-                            (CrefSyntax)SyntaxFactory.TypeCref(typeSyntax) :
-                            SyntaxFactory.NameMemberCref(typeSyntax);
-                        crefSymbols = BindCref(crefWrapper, binder);
-                    }
+                    var crefWrapper = typeSyntax is PredefinedTypeSyntax ?
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: :bulb: This would be faster with `typeSyntax.IsKind(SyntaxKind.PredefinedType)`

Comment (alrz) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Sure. should I consider this for like *all* type checks? there's a plenty in this file but it seems like that none of them are hot paths.

**Comment by alrz** [comment]
@jcouv Anything else need to be done here?

**Comment by jaredpar** [comment]
CC @roslyn-compiler for one more review.

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 8]
```diff
@@ -1759,94 +1751,92 @@ private bool IsInTypeofExpression(int position)
             BoundNode boundNodeForSyntacticParent,
             Binder binderOpt)
         {
-            var boundExpr = lowestBoundNode as BoundExpression;
-            var highestBoundExpr = highestBoundNode as BoundExpression;
-            if (boundExpr != null)
+            if (!(lowestBoundNode is BoundExpression boundExpr))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >if (!(lowestBoundNode is BoundExpression boundExpr)) [](start = 12, length = 52)

It feels like there is too much churn in this function.
